webapp/src/ts/services/performance.service.ts

```diff
@@ -0,0 +1,92 @@
+import { Inject, Injectable } from '@angular/core';
+import { DOCUMENT } from '@angular/common';
+
+import { TelemetryService } from '@mm-services/telemetry.service';
+import { AuthService } from '@mm-services/auth.service';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class PerformanceService {
+  private trackPerformance = false;
+  private readonly CAN_TRACK_PERFORMANCE = 'track_performance';
+  private readonly PERFORMANCE_PREFIX = 'perf:';
+  private readonly APDEX_SUBFIX = ':apdex';
+  private readonly APDEX_SATISFIED = ':satisfied';
+  private readonly APDEX_TOLERATING = ':tolerable';
+  private readonly APDEX_FRUSTRATED = ':frustrated';
+  private readonly APDEX_AGGREGATE = ':aggregate';
+  private readonly APDEX_T = 3 * 1000;
+  private readonly APDEX_TOLERANCE = 4; // 4xT
+
+  constructor(
+    private telemetryService: TelemetryService,
+    private authService: AuthService,
+    @Inject(DOCUMENT) private document: Document,
+  ) {
+    this.authService
+      .has(this.CAN_TRACK_PERFORMANCE)
+      .then(result => this.trackPerformance = result);
+  }
+
+  track(name: string) {
+    if (!this.trackPerformance || !name || !this.document?.defaultView) {
+      return;
+    }
+
+    const startTime = this.document.defaultView.performance.now();
+    return {
+      setName: newName => name = newName,
+      stop: (recordApdex = false, prefix = true) => this.recordPerformance(name, startTime, recordApdex, prefix),
+    };
+  }
+
+  /**
+   * Records Telemetry entry
+   * @param name        Telemetry entry's name
+   * @param startTime   Process start time in milliseconds
+   * @param recordApdex If true, then record Apdex as additional Telemetry entry
+   * @param prefix      If false, then don't prefix with "perf:" the Telemetry entry's name. Introduced to keep old
+   *                    Telemetry entries' name.
+   * @private
+   */
+  private async recordPerformance(name: string, startTime: number, recordApdex = false, prefix = true) {
+    if (!this.trackPerformance || !this.document?.defaultView) {
+      return;
+    }
+
+    const telemetryName = prefix ? this.PERFORMANCE_PREFIX + name : name;
+    const time = this.document.defaultView.performance.now() - startTime;
+    await this.telemetryService.record(telemetryName, time);
+
+    if (recordApdex) {
+      const { component, aggregate } = this.getApdexLabels(name, time);
+      await this.telemetryService.record(component, time);
+      await this.telemetryService.record(aggregate, time);
```

Comment:
I don't think we should duplicate this because it doubles the number of records that need to be stored and increases the docs to sync. We can easily aggregate them on the server if we record the `component`.

Reply:
> it doubles the number of records that need to be stored and increases the docs to sync

There was a back-and-forth on the aggregate, but it can also be calculated when building the reports